pandas/tseries/offsets.py

```diff
@@ -1155,7 +1155,12 @@ def apply_index(self, i):
         shifted = liboffsets.shift_months(i.asi8, self.n, self._day_opt)
         # TODO: going through __new__ raises on call to _validate_frequency;
         #  are we passing incorrect freq?
-        return type(i)._simple_new(shifted, freq=i.freq, dtype=i.dtype)
+        dta = i
+        if not isinstance(i._data, np.ndarray):
```

Comment:
WAT?

Reply:
compat for getting either a DTA or a DTI, we already do this in a couple places.  I think we can do it just once by putting it in the `apply_index_wraps` decorator